### PR TITLE
Popover: anchor correctly to parent node when no explicit anchor is passed

### PR DESCRIPTION
**Pull request by ciampo**
## What?

This PR fixes a regression introduced by #40740 which caused the popover to position incorrectly when no explicit anchor is passed via props. In this case, the popover is supposed to use its parent node as the anchor — as per the README:

https://github.com/WordPress/gutenberg/blob/8f0c091c92ff28279d55844bca6b941280e9509a/packages/components/src/popover/README.md?plain=1#L3

## Why?

Fixing a regression. Here's how the code used to look like before #40740:

https://github.com/WordPress/gutenberg/blob/b99e6688fba7037ccee46cb6b7d70917a5f2c418/packages/components/src/popover/index.js#L123-L129

## How?

The fix is simple — instead of using `fallbackAnchor.current` to compute the anchor positioning, we use its parent node (`fallbackAnchor.current.parentNode`).

This change also required a slight tweak with the styles of the arrow's triangle which somehow seemed to conflict with some floating ui's calculation (switched from nested `absolute` positioning to a simpler `display: block`)

### Storybook

- Play around with the different `placement` options in the default Storybook example, and notice how the popover positions itself correctly with respect to the toggle button.
- Repeat the same on `trunk` and notice how the popover anchors itself around a single pixel on screen, instead of the full button.
- Check that there are no regressions in components using `Popover` (e.g. `Tooltip`, `Dropdown`...)

### Editor

**Please do test thoroughly in the editor, as I wasn't always able to "find" all of the popovers listed below while browsing the app in the browser.**

<details>

<summary> This PR should introduce bug fixes (or at least, not cause regressions) for all instances of `Popover` that don't pass an explicit anchor via props. This is what I've found when searching in the codebase: (click to expand)</summary>

- https://github.com/WordPress/gutenberg/blob/5c1e9139c75b72cfd04af15da3a1b5affcc250cf/packages/block-editor/src/components/block-switcher/pattern-transformations-menu.js#L59-L62
- https://github.com/WordPress/gutenberg/blob/d9d06f332d9c5057a19f8ad334625bcbce3e137a/packages/block-editor/src/components/block-switcher/preview-block-popover.js#L16-L20
- https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/block-editor/src/components/url-input/index.js#L542
- https://github.com/WordPress/gutenberg/blob/33a9c48fc19b25b45636997e906f04bc06d6cc7e/packages/block-editor/src/components/url-popover/index.js#L32-L38
- https://github.com/WordPress/gutenberg/blob/8b1492131355aa6d42272faba25b841f0cd4509d/packages/nux/src/components/dot-tip/index.js
- https://github.com/WordPress/gutenberg/blob/8b1492131355aa6d42272faba25b841f0cd4509d/packages/widgets/src/blocks/legacy-widget/edit/form.js#L104-L108

`Popover` is also used without an anchor in the `PaletteEdit`, but since the component doesn't have a Storybook example, we'd have to check directly in the editor:
- https://github.com/WordPress/gutenberg/blob/a5ef504432e34afc0792eae62e44d71590c031eb/packages/edit-site/src/components/global-styles/color-palette-panel.js#L49-L76
- https://github.com/WordPress/gutenberg/blob/2224d4f59892aec92718c0e9a48866d28967156d/packages/edit-site/src/components/global-styles/gradients-palette-panel.js#L55-L82

</details>

## Screenshots or screencast 

On `trunk`:

https://user-images.githubusercontent.com/1083581/182825267-e23ba4e2-1367-488e-a0d0-c51f9d5f6358.mp4

On this PR:

https://user-images.githubusercontent.com/1083581/182825039-e8ea5647-88de-4940-a71e-d837a96bbd13.mp4

Editing color palette:

| `trunk` | This PR |
|---|---|
| Screenshot 2022-08-04 at 13 02 34 | Screenshot 2022-08-04 at 13 04 07 |

### CODE REVIEW

File: packages/block-editor/src/components/url-popover/image-url-input-ui.js
Comment (ciampo) [inline packages/block-editor/src/components/url-popover/image-url-input-ui.js:253]: Passing an explicit anchorRef here so that the URL popover is anchored to the button (and not to the whole toolbar).

File: packages/components/src/popover/index.js
Comment (aaronrobertshaw) [inline packages/components/src/popover/index.js]:
```suggestion
			// If no explicit ref is passed via props, fall back to
```

Typo?

**Review by talldan — COMMENTED** [review]
Nice fix. I couldn't spot any issues.

I guess it just needs a changelog update.

**Review by talldan — APPROVED** [review]

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for working on this @ciampo 👍 

In general, this tests well for me. I didn't find any instances of Popovers without an explicit anchor than those listed in the testing instructions.

I wasn't sure how to test the `dop-tip` popover but was able to eyeball the others on trunk and this PR. 

✅ Pattern transformations menu - No change.
✅ Preview block popover - No change.
✅ URL Input - No change.
❓ URL Popover - Popover is slightly to the left and doesn't align to the block toolbar now. I'd say this is ok but wanted to note it at least.
❓ Legacy widget - Popover position is better but could do with constraining to viewport. Maybe a follow-up?
✅ Color palette panel - Positioning is improved
✅ Gradient palette panel - Positioning is improved

<details>
<summary>URL Popover Screenshots</summary>

| Before | After |
|---|---|
| Screen Shot 2022-08-09 at 3 10 50 pm | Screen Shot 2022-08-09 at 5 19 04 pm | 
</details>

<details>
<summary>Legacy Widget Popover Videos</summary>

| Before | After |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/183592097-30eae198-1b09-4678-a586-ecc93c6ef826.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/183592360-cfa23349-1212-4336-9195-b7ae28f1fab0.mp4" /> |

</details>

I don't think these small issues should block this PR at all. Also, a changelog would be handy for this one.

**Comment by talldan** [comment]
> Legacy widget - Popover position is better but could do with constraining to viewport. Maybe a follow-up?

True, it should resize to the height of the viewport. This PR is ten times better than trunk though. I can follow up with a fix for that once this is merged (I just about remember how it should work).

File: packages/components/src/popover/index.js
Comment (ciampo) [reply packages/components/src/popover/index.js]: 😅 thank you , done in https://github.com/WordPress/gutenberg/pull/42971/commits/83343964a39e303cc545dc9a7fb498e206478f39

**Comment by ciampo** [comment]
Thank you both for the reviews! I've fixed the typo and added a CHANGELOG entry.

> URL Popover - Popover is slightly to the left and doesn't align to the block toolbar now. I'd say this is ok but wanted to note it at least.

That makes sense, and it's caused by the fact that I've decided to [pass an explicit anchor ref](https://github.com/WordPress/gutenberg/pull/42971#discussion_r937699009) to the `<URLPopover />` — that causes the popover to be centre-aligned with its toolbar button.

I can also drop the changes to `packages/block-editor/src/components/url-popover/image-url-input-ui.js` completely from this PR, which means that the popover will be centre-aligned with respect to the whole group of icons — let me know!

I think we would still need to polish the popover's placement anyway top align it with the rest of the popovers that appear when clicking the other toolbar buttons (they all seem to be `bottom-start` aligned, and with some vertical `offset` between the toolbar and the popover) — but since I don't have much knowledge about all the possible scenarios in which `URLPopover` is used, I'd rather have someone with more experience tweak it in a follow-up.

> > Legacy widget - Popover position is better but could do with constraining to viewport. Maybe a follow-up?
> 
> True, it should resize to the height of the viewport. This PR is ten times better than trunk though. I can follow up with a fix for that once this is merged (I just about remember how it should work).

Sounds good to me! Enabling `shift`-ing behavior may just be enough 🤞  thank you @talldan !

---

I'm assuming we're good to merge one CI is green ✅  and open follow-up PRs where needed as discussed above

**Comment by andrewserong** [comment]
Thanks so much for fixing this one up @ciampo, I think this explains (and resolves) a bunch of the confusing issues we've had since the refactor, so it's a great step forward.

**Comment by talldan** [comment]
The legacy widget popover is still on my radar. I think it will require using floating-ui's `size` middelware, but this is a popover also has `__unstableForcePosition` set to `true` which deactivates `size`.

It seems like the props of `Popover` might need a little refactoring. I think `__unstableForcePosition` could be removed in favor of separate `resize` and `flip` props, but it will require a refactor PR first and then the bug fix. A job for next week 😄

**Comment by ciampo** [comment]
> It seems like the props of `Popover` might need a little refactoring. I think `__unstableForcePosition` could be removed in favor of separate `resize` and `flip` props, but it will require a refactor PR first and then the bug fix. A job for next week 😄

Absolutely! Opened #43191 to keep track of this separately (and in case we need to continue the conversation before opening a PR)

**Pull request merged**
